### PR TITLE
[ci] switch to manylinux_2_28 for Linux artifacts (fixes #5514, fixes #5589)

### LINKED ISSUE #5514: Use manylinux image to produce wheels
## Summary

Created from https://github.com/microsoft/LightGBM/issues/5390#issuecomment-1204486025.

The following changes should be made in LightGBM:

* move from supporting `GLIBC >= 2.14` to `GLIBC >= 2.23`
* start producing Python wheels using one of the `manylinux` images (https://github.com/pypa/manylinux)
    - this implies ending official support for Ubuntu 14.04 in LightGBM

## Motivation

It has been really difficult to maintain support for older operating systems and versions of GLIBC.

References:

* https://github.com/microsoft/LightGBM/issues/5390#issuecomment-1203495465
* https://github.com/microsoft/LightGBM/pull/4893#issuecomment-1039204806
* https://github.com/microsoft/LightGBM/pull/4893#issuecomment-1039812782

Moving to newer base image and higher floor for GLIBC and supported Ubuntu version should help to alleviate some maintenance burden.

## Description

As described in https://github.com/microsoft/LightGBM/issues/5390#issuecomment-1204486025, LightGBM should upgrade to a `manylinux` base image supporting at least `GLIBC==2.23` (and not too much higher) for producing wheels.

The `manylinux` README shows the following compatibility for wheels produced on those images (https://github.com/pypa/manylinux).

image

https://ubuntu.com/about/release-cycle describes the release cycle for versions of Ubuntu

image

Using an approach like the following

```shell
docker run \
    --rm \
    --entrypoint="" \
    -it quay.io/pypa/manylinux1_x86_64 \
    ldd --version
```

I found that the GLIBC version for the published `manylinux` images is as follows.

| image                    | GLIBC |
|:-----------------:|:------:|
|manylinux1            | 2.5      |
|manylinux2010     | 2.12    |
|manylinux2014     | 2.17    |
|manylinux_2_28   | 2.28   |

## References

Other related conversations:

* discussion from 2020 about possibly using `manylinux_2010`: https://github.com/microsoft/LightGBM/pull/3421#issuecomment-711559332
* plan to migrate to PEP600 `manylinux_x_y` tags eventually: https://github.com/microsoft/LightGBM/issues/4484#issuecomment-889045289

### LINKED ISSUE #5589: [ci] compilation failing on macOS jobs using gcc
## Description

All CI jobs on macOS started failing some time in the last 12 hours, with the following error.

```text
[100%] Linking CXX shared library /Users/runner/work/LightGBM/LightGBM/lib_lightgbm.so
0  0x10a416ffa  __assert_rtn + 139
1  0x10a24a28d  mach_o::relocatable::Parser<x86_64>::parse(mach_o::relocatable::ParserOptions const&) + 4989
2  0x10a23af8f  mach_o::relocatable::Parser<x86_64>::parse(unsigned char const*, unsigned long long, char const*, long, ld::File::Ordinal, mach_o::relocatable::ParserOptions const&) + 207
3  0x10a2b19d4  ld::tool::InputFiles::makeFile(Options::FileInfo const&, bool) + 2036
4  0x10a2b4fa0  ___ZN2ld4tool10InputFilesC2ER7Options_block_invoke + 48
5  0x7ff81923d34a  _dispatch_client_callout2 + 8
6  0x7ff81924e8f5  _dispatch_apply_invoke + 213
7  0x7ff81923d317  _dispatch_client_callout + 8
8  0x7ff81924cc0c  _dispatch_root_queue_drain + 673
9  0x7ff81924d25c  _dispatch_worker_thread2 + 160
10  0x7ff8193f0f8a  _pthread_wqthread + 256
A linker snapshot was created at:
	/tmp/lib_lightgbm.so-2022-11-15-050227.ld-snapshot
ld: Assertion failed: (_file->_atomsArrayCount == computedAtomCount && "more atoms allocated than expected"), function parse, file macho_relocatable_file.cpp, line 2061.
collect2: error: ld returned 1 exit status
make[3]: *** [/Users/runner/work/LightGBM/LightGBM/lib_lightgbm.so] Error 1
make[2]: *** [CMakeFiles/_lightgbm.dir/all] Error 2
make[1]: *** [CMakeFiles/_lightgbm.dir/rule] Error 2
make: *** [_lightgbm] Error 2
```

This occurs after compiling all objects, at the stage where the linker creates `lib_lightgbm.so`.

## Reproducible example

For example, on #5588, all Python jobs on macOS are failing ([link](https://github.com/microsoft/LightGBM/actions/runs/3467557216/jobs/5792537483))

image

And all the two R-package jobs on macOS using `gcc`.

image

## Additional Comments

It seems like the issue might be "the version of `ld` included with XCode 14.0 breaks compatibility with `gcc`": https://github.com/iains/gcc-12-branch/issues/6.

Other related discussions:

* https://github.com/orgs/Homebrew/discussions/3659
* https://github.com/open-source-parsers/jsoncpp/issues/1434
* https://stackoverflow.com/a/73765819/3986677
* https://developer.apple.com/forums/thread/714565

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #5514.
Fixes #5589.
Contributes to #5390.
Contributes to #5391 (via https://github.com/microsoft/LightGBM/pull/5396#issuecomment-1203168124).

Proposes switching from an Ubuntu 14.04 image to one based on `pypa/manylinux_2_28` for building artifacts on Linux including:

* Python source and built distributions (e.g. wheels)
* `lightgbmlib.jar` (Java interface to LightGBM)
* `lib_lightgbm.so` (C++ shared library)

And increases the floors on symbol versions for key linked libraries on Linux.

* `GLIBC`: `<= 2.14` ([October 2011](https://sourceware.org/glibc/wiki/Glibc%20Timeline)) to `<= 2.28` ([August 2018](https://sourceware.org/glibc/wiki/Glibc%20Timeline))
* `GLIBCXX`: `<= 3.4.19` to `<= 3.4.22`
* `GOMP`: `== 1.0` ([January 2006](https://github.com/gcc-mirror/gcc/blame/0cbb756fe9c8e13a743bdcf599481b31bd01ddbc/libgomp/libgomp.map#L243)) to `<= 4.5` ([October 2015](https://github.com/gcc-mirror/gcc/blame/0cbb756fe9c8e13a743bdcf599481b31bd01ddbc/libgomp/libgomp.map#L343))

### Notes for Reviewers

This PR uses images build from https://github.com/guolinke/lightgbm-ci-docker/pull/27. That `lightgbm-ci-docker` PR should be reviewed and merged first.

### Why all these `if-else` forks in the CI scripts?

`pypa/manylinux_2_28` is based on Alma Linux, so it used `dnf` and `yum` as package managers (where Ubuntu uses `apt`).

### CODE REVIEW

**Comment by jameslamb** [comment]
@guolinke @shiyu1994 @jmoralez @StrikerRUS I have been struggling to fix the issues showing up on the Ubuntu 14.04 image (https://github.com/microsoft/LightGBM/pull/5588#issuecomment-1319516599).

Instead of spending more time on that, given that we want to move to a new image for building wheels anyway (#5514), can we please just focus on making that switch?

What I'm asking for from you:

* read the plan at https://github.com/microsoft/LightGBM/issues/5153#issuecomment-1319532263 and tell me if you disagree with any part of it
* review https://github.com/guolinke/lightgbm-ci-docker/pull/27 and this PR together
    - (@guolinke , I think we should add @shiyu1994 as a maintainer and @jmoralez as a contributor there, so they can review)

File: .ci/setup.sh
Comment (jameslamb) [inline .ci/setup.sh:10]: this fixes #5589

Comment (ARILDMT) [reply .ci/setup.sh:10]: Com

**Comment by jameslamb** [comment]
✅ building successfully with `COMPILER=clang` ([commit](https://github.com/microsoft/LightGBM/pull/5580/commits/7a59c9bed7114b19350b532dd63f680f0c713ff6) - [build](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13868))

✅ building successfully with `COMPILER=gcc` ([commit](https://github.com/microsoft/LightGBM/pull/5580/commits/9ea23ee02abccd2d3313192e2d419aedf9e9ae58) - [build](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13867&view=results))

**Comment by jameslamb** [comment]
@guolinke whenever you have time, could you review this PR next?

This will unblock LightGBM's CI and allow us to continue making progress on the others things needed to stabilize the CI and improve the process of building Python wheels.

File: python-package/README.rst
Comment (guolinke) [inline python-package/README.rst:29]: I think this is for version > 3.3.3, should we add a note about the minimal version of glibc for version <= 3.3.3 ?

**Review by guolinke — APPROVED** [review]

**Comment by guolinke** [comment]
A minor comment about the documentation.

File: python-package/README.rst
Comment (jameslamb) [reply python-package/README.rst]: Sure!

I added this note inline in https://github.com/microsoft/LightGBM/pull/5580/commits/8def665466fd83ce7c9c7a61b586509c5c2f2db0.

The next time we change this floor, it should probably be replaced with a table similar to those in the description of #5514.

**Comment by jameslamb** [comment]
Thanks very much @guolinke ! I'll merge this without waiting on any other reviews so we can keep making progress on stabilizing CI and preparing for the v4.0.0 release.

**Pull request merged**
